controller/internal/enforcer/nfqdatapath/dns.go

```diff
@@ -0,0 +1,133 @@
+package nfqdatapath
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.aporeto.io/trireme-lib/controller/pkg/packet"
+	"go.aporeto.io/trireme-lib/policy"
+	"go.uber.org/zap"
+)
+
+type serveDNS struct {
+	*Datapath
+}
+
+func socketOptions(_, _ string, c syscall.RawConn) error {
+	var opErr error
+	err := c.Control(func(fd uintptr) {
+		if err := syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt); err != nil {
+			zap.L().Error("Failed to mark connection", zap.Error(err))
+		}
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return opErr
+}
+
+func listenUDP(network, addr string) (net.PacketConn, error) {
+	var lc net.ListenConfig
+
+	lc.Control = socketOptions
+
+	return lc.ListenPacket(context.Background(), network, addr)
+}
+
+func forwardDNSReq(r *dns.Msg, ip net.IP, port uint16) (*dns.Msg, []string, error) {
+	var ips []string
+	c := new(dns.Client)
+	c.Dialer = &net.Dialer{
+		Control: func(_, _ string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
+				if err := syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt); err != nil {
+					zap.L().Error("Failed to assing mark to socket", zap.Error(err))
+				}
+			})
+		},
+		Timeout: 500 * time.Millisecond,
+	}
+
+	in, _, err := c.Exchange(r, net.JoinHostPort(ip.String(), strconv.Itoa(int(port))))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, ans := range in.Answer {
+		if ans.Header().Rrtype == dns.TypeA {
+			t, _ := ans.(*dns.A)
+			ips = append(ips, t.A.String())
+		}
+
+		if ans.Header().Rrtype == dns.TypeAAAA {
+			t, _ := ans.(*dns.AAAA)
+			ips = append(ips, t.AAAA.String())
+		}
+	}
+
```

Comment:
Entries have TTL. I think we should use the TTL to expire them. There is some interesting logic we can follow on that.

Reply:
Yes, I will be doing TTL in a different PR. We will also have to add new api's in our ApplicationACLS to remove entries as well on expiry.